OSDOCS-5202re:Documented vSphere config provisioning of nodes with sta (#62677)
openshift/openshift-docs: modules/nodes-vsphere-machine-set-scaling-static-ip.adoc

Reviewer: I'm wondering if this file should be broken up into a CONCEPT and a PROCEDURE. Seems like there's both types of information in here and it is rather long. See https://redhat-documentation.github.io/modular-docs/

Author: HI @skrthomas . I created a new module but I'm not sure if "How to" is a way to lead such a concept module? I'm struggling with this one as the procedure module has a very specific title with 3 key nouns being emphasized to really filter down the specific use case.

Reviewer: I made a couple of suggestions for your titles.

// Module included in the following assemblies:
//
// * post_installation_configuration/node-tasks.adoc

:_content-type: PROCEDURE
[id="nodes-vsphere-machine-set-scaling-static-ip_{context}"]

> = Using a machine set to scale machines with configured static IP addresses

Reviewer: ```suggestion
= Configuring machine set scaling of machines with static IP addresses
```

Author: I'll retain the original title as, IMO,  it's clearer

+
[source, terminal]
----
$ oc create -f ipaddress.yaml
----
+
The following example shows an `IPAddress` resource with defined network configuration information and one defined static IP address:
+
[source,yaml]
----
apiVersion: ipam.cluster.x-k8s.io/v1alpha1
kind: IPAddress
metadata:
  name: cluster-dev-9n5wg-worker-0-m7529-ipaddress-0-0
  namespace: openshift-machine-api
spec:
  address: 192.168.204.129
  claimRef: <1>
    name: cluster-dev-9n5wg-worker-0-m7529-claim-0-0
  gateway: 192.168.204.1
  poolRef: <2>
    apiGroup: ipamcontroller.example.io
    kind: IPPool
    name: static-ci-pool
  prefix: 23
----
<1> The name of the target `IPAddressClaim` resource. 
<2> Details information about the static IP address or addresses from your nodes. 
+
[NOTE]
====
By default, the external controller automatically scans any resources in the machine set for recognizable address pool types. When the external controller finds `kind: IPPool` defined in the `IPAddress` resource, the controller binds any static IP addresses to the `IPAddressClaim` resource.
====

. Update the `IPAddressClaim` status with a reference to the `IPAddress` resource:
+
[source, terminal]
----

> $ oc --type=merge patch IPAddressClaim cluster-dev-9n5wg-worker-0-m7529-claim-0-0 -p='{"status":{"addressRef": {"name": "cluster-dev-9n5wg-worker-0-m7529-ipaddress-0-0}}}' -n openshift-machine-api --subresource=status

Reviewer 2: We missed a `"` here
`"cluster-dev-9n5wg-worker-0-m7529-ipaddress-0-0` 
shoud change to 
`"cluster-dev-9n5wg-worker-0-m7529-ipaddress-0-0"`

Reviewer 2: Others lgtm

  name: <infrastructure_id>-<role>
  namespace: openshift-machine-api
spec:
  replicas: 0
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-cluster: <infrastructure_id>
      machine.openshift.io/cluster-api-machineset: <infrastructure_id>-<role>
  template:
    metadata:
      labels:
        ipam: "true"
        machine.openshift.io/cluster-api-cluster: <infrastructure_id>
        machine.openshift.io/cluster-api-machine-role: worker
        machine.openshift.io/cluster-api-machine-type: worker
        machine.openshift.io/cluster-api-machineset: <infrastructure_id>-<role>
    spec:
      lifecycleHooks: {}
      metadata: {}
      providerSpec:
        value:
          apiVersion: machine.openshift.io/v1beta1
          credentialsSecret:
            name: vsphere-cloud-credentials
          diskGiB: 120
          kind: VSphereMachineProviderSpec
          memoryMiB: 8192
          metadata: {}
          network:
            devices:
            - addressesFromPools: <1>
              - group: ipamcontroller.openshift.io
                name: static-ci-pool
                resource: IPPool
              nameservers:

>               - "192.168.204.1" <2>

Author: SME query: Internal [writer guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#ip-addresses) state that only a certain range of IP addresses should be used in public-facing docs (depends on whether address are private or public). Are the addresses in the example private addresses?

![Screenshot from 2023-07-18 12-06-36](https://github.com/openshift/openshift-docs/assets/57954076/17fcdd63-ed0a-4de7-bcf8-8eda71647d36)

Reviewer 3: These are all private IP addresses.  That said, if they need to be changed per guidelines I have no issue with that.

Author: Thanks: `192.168.204.1` falls into the acceptable range according to the above guidance.

    spec:
      lifecycleHooks: {}
      metadata: {}
      providerSpec:
        value:
          apiVersion: machine.openshift.io/v1beta1
          credentialsSecret:
            name: vsphere-cloud-credentials
          diskGiB: 120
          kind: VSphereMachineProviderSpec
          memoryMiB: 8192
          metadata: {}
          network:
            devices:
            - addressesFromPools: <1>
              - group: ipamcontroller.openshift.io
                name: static-ci-pool
                resource: IPPool
              nameservers:
              - "192.168.204.1" <2>
              networkName: qe-segment-204
          numCPUs: 4
          numCoresPerSocket: 2
          snapshot: ""
          template: rvanderp4-dev-9n5wg-rhcos-generated-region-generated-zone
          userDataSecret:
            name: worker-user-data
          workspace:
            datacenter: IBMCdatacenter
            datastore: /IBMCdatacenter/datastore/vsanDatastore
            folder: /IBMCdatacenter/vm/rvanderp4-dev-9n5wg
            resourcePool: /IBMCdatacenter/host/IBMCcluster//Resources
            server: vcenter.ibmc.devcluster.openshift.com
----

> <1> Specifies an IP pool, which lists a static IP address or a range of static IP addresses. The IP Pool can either be a reference to a custom resource definition (CRD) or a resource supported by the `IPAddressClaims` resource handler. The machine controller accesses static IP addresses listed in the machine set's configuration and then allocates each address to each machine. 

Reviewer 2: I have a question here. 
`The IP Pool can either be a reference to a custom resource definition (CRD) or a resource supported by the `IPAddressClaims` resource handler.`
I see we mentioned `ipamcontroller.openshift.io` in this doc PR, and looks like it's defined in custom resource definition (CRD) [here](custom resource definition (CRD)) as an sample, so should we also add it as an IPPool sample to doc? 
Or shall we provide a example link of resource supported by the `IPAddressClaims` resource handler?

@rvanderp3 Could you give some help for it? Thanks.

Author: Thanks, @rvanderp3 and @sgaoshang . I swapped out three instances of `ipamcontroller.openshift.io` and replaced them with `ipamcontroller.example.io`.

Is the following description still OK?

> Specifies an IP pool, which lists a static IP address or a range of static IP addresses. The IP Pool can either be a reference to a custom resource definition (CRD) or a resource supported by the `IPAddressClaims` resource handler. The machine controller accesses static IP addresses listed in the machine set's configuration and then allocates each address to each machine.

Reviewer 2: Thanks @rvanderp3 , could you pls tell how to set IPPool range in the IPAddressClaim?

----

. Create an `IPAddress` resource. The following example shows an `IPAddress` resource with defined network configuration information and one defined static IP address:
+
[source,yaml]
----
apiVersion: ipam.cluster.x-k8s.io/v1alpha1
kind: IPAddress
metadata:
  name: cluster-dev-9n5wg-worker-0-m7529-ipaddress-0-0
  namespace: openshift-machine-api
spec:
  address: 192.168.204.129
  claimRef: <1>
    name: cluster-dev-9n5wg-worker-0-m7529-claim-0-0
  gateway: 192.168.204.1
  poolRef: <2>
    apiGroup: ipamcontroller.openshift.io
    kind: IPPool
    name: static-ci-pool
  prefix: 23
----
<1> The name of the target `IPAddressClaim` resource. 
<2> Details information about the static IP address or addresses from your nodes. 
+
[NOTE]
====
By default, the external controller automatically scans any resources in the machine set for recognizable address pool types. When the external controller finds `kind: IPPool` defined in the `IPAddress` resource, the controller binds any static IP addresses to the `IPAddressClaim` resource.
====

. Optional: To manually perform a binding operation, enter the following command:
+
[source, terminal]
----

> $ oc create -f ipaddress.yaml

Author: Thanks, @rvanderp3 . Updated to state:

![Screenshot from 2023-09-14 10-45-14](https://github.com/openshift/openshift-docs/assets/57954076/48537c0a-c769-498f-849a-2e9c1d9eda24)

//
// * post_installation_configuration/node-tasks.adoc

:_content-type: PROCEDURE
[id="nodes-vsphere-machine-set-scaling-static-ip_{context}"]
= Using a machine set to scale machines with configured static IP addresses

You can use a machine set to scale machines with configured static IP addresses. 

:FeatureName: Static IP addresses for vSphere nodes
include::snippets/technology-preview.adoc[leveloffset=+1]

After you configure a machine set to request a static IP address for a machine, the machine controller creates an `IPAddressClaim` resource in the `openshift-machine-api` namespace. The external controller then creates an `IPAddress` resource and binds any static IP addresses to the `IPAddressClaim` resource.

[IMPORTANT]
====
Your organization might use numerous types of IP address management (IPAM) services. If you want to enable a particular IPAM service on {product-title}, you might need to manually create the `IPAddressClaim` resource in a YAML definition and then bind a static IP address to this resource by entering the following command in your `oc` CLI:

[source, terminal]
----
oc create -f <ipaddressclaim_filename>
----

> See https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#nw-multus-ipam-object_configuring-additional-network 

Author: SME query: Worthwhile linking to this section? 

https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/configuring-additional-network.html#nw-multus-ipam-object_configuring-additional-network

Reviewer 3: While IPAM is used in both scenarios, how it is used and implemented is quite different.  I don't think we should link to this document.

Author: Thanks. Removed the link.

Author: Removed link.

          userDataSecret:
            name: worker-user-data
          workspace:
            datacenter: IBMCdatacenter
            datastore: /IBMCdatacenter/datastore/vsanDatastore
            folder: /IBMCdatacenter/vm/rvanderp4-dev-9n5wg
            resourcePool: /IBMCdatacenter/host/IBMCcluster//Resources
            server: vcenter.ibmc.devcluster.openshift.com
----
<1> Specifies an IP pool, which lists a static IP address or a range of static IP addresses. The IP Pool can either be a reference to a custom resource definition (CRD) or a resource supported by the `IPAddressClaims` resource handler. The machine controller accesses static IP addresses listed in the machine set's configuration and then allocates each address to each machine. 
<2> Lists a nameserver. You must specify a nameserver for nodes that receive static IP address, because the Dynamic Host Configuration Protocol (DHCP) network configuration does not support static IP addresses. 

. Scale the machine set by entering the following commands in your `oc` CLI:
+
[source, terminal]
----
$ oc scale --replicas=2 machineset <machineset> -n openshift-machine-api
----
+

> Or:

Author: SNE query: I'm a bit confused with the steps for this procedure. I might have stumbled into using a mixture of manual steps and machine controller automated steps? WDYT? 

Maybe I should state both routes? Or does a blend exist?

I mentioned the external controller as I seen it referenced in [this GH page](https://github.com/openshift/enhancements/blob/5d8751c23dc39a878b3b73d64fe44745aceb25d9/enhancements/network/static-ip-addresses-vsphere.md#scaling-new-nodes-with-machinesets).

Reviewer 3: It is very much a blend.  The machine API controller creates a resource that either some external controller(i.e. the GH page you linked) or an admin handles via `oc` commands.  My intention was to demonstrate how this can be done with `oc` so that an administrator can create the necessary automation based on what they want or need to use.

Reviewer 3: I would keep it around as a note.

Author: Thanks, @rvanderp3 . I converted the paragraph to a note and prepended it with "By default". Having a string of notes can be distracting from a docs perspective, so I converted the other note to an optional step. WDYT?

![Screenshot from 2023-08-15 12-09-02](https://github.com/openshift/openshift-docs/assets/57954076/b581afb5-237f-4832-9ad1-257c62c6d73e)
